<h1><p style="color: green">Conclusion</p></h1>


In [1]:
from prettytable import PrettyTable


In [15]:
ptable1 = PrettyTable()
ptable1.title = " Model Comparision "
ptable1.field_names = ['Model Name','Hyperparameter Tunning', 'Accuracy']

print("\n\n ********** Machine Learning Model Comparison ************")
ptable1.add_row(["Support Vector Machine (poly)", "Done", "96.3%"])
ptable1.add_row(["Logistic Regression", "Done", "95.86%"])
ptable1.add_row(["XGBoost", "Done", "95.05%"])
ptable1.add_row(["Random Forest", "Done", "91.96%"])
ptable1.add_row(["K-Nearest Neighbours", "Done", "90.23%"])
ptable1.add_row(["DecisionTree", "Done", "86.39%"])

print(ptable1)

ptable5 = PrettyTable()
ptable5.title = "Permutation Importance of Features:"
ptable5.field_names = ['Feature', 'Importance']
print("\n\n ********** Feature Importances For Best Model  ************")

ptable5.add_row(["tBodyGyroJerk-entropy()-X", "0.081778"])
ptable5.add_row(["fBodyGyro-entropy()-X", "0.051239"])
ptable5.add_row(["fBodyAcc-entropy()-X", "0.016966"])
ptable5.add_row(["tGravityAcc-mean()-Y", "0.009841"])
ptable5.add_row(["fBodyBodyAccJerkMag-entropy()", "0.009026"])
ptable5.add_row(["tGravityAcc-min()-Y", "0.008415"])
ptable5.add_row(["tGravityAcc-max()-Y", "0.007737"])
ptable5.add_row(["tBodyGyroJerk-entropy()-Y", "0.005904"])
ptable5.add_row(["fBodyGyro-skewness()-X", "0.005836"])

print(ptable5)

ptable2 = PrettyTable()
ptable2.title = " Model Comparision "
ptable2.field_names = ['Model Name','Hyperparameter Tunning', 'Accuracy']

print("\n\n*** ML Model Comparison after Dimensionality Reduction using PCA ***")
ptable2.add_row(["Support Vector Machine (rbf)", "Done", "95.76%"])
ptable2.add_row(["Logistic Regression", "Done", "94.13%"])
ptable2.add_row(["XGBoost", "Done", "92.57%"])
ptable2.add_row(["Random Forest", "Done", "91.21%"])
ptable2.add_row(["K-Nearest Neighbors", "Done", "87.89%"])

print(ptable2)

ptable3 = PrettyTable()
ptable3.title = " Model Comparision "
ptable3.field_names = ['Model Name','Hyperparameter Tunning', 'Accuracy']

print("\n\n** ML Model Comparison after Dimensionality Reduction using KBest **")
ptable3.add_row(["Support Vector Machine (rbf)", "Done", "87.00%"])
ptable3.add_row(["Logistic Regression", "Done", "86.39%"])
ptable3.add_row(["Random Forest", "Done", "85.14%"])
ptable3.add_row(["XGBoost", "Done", "83.41%"])
ptable3.add_row(["K-Nearest Neighbors", "Done", "82.25%"])

print(ptable3)

ptable4 = PrettyTable()
ptable4.title = " Model Comparision "
ptable4.field_names = ['Model Name','Hyperparameter Tunning', 'categorical_crossentropy', 'Accuracy']

print("\n\n ********************************* Deep Learning LSTM Model Comparision ***********************************")
ptable4.add_row(["LSTM With 1_Layer(neurons:32)","Done","0.50", "0.90"])
ptable4.add_row(["LSTM With 2_Layer(neurons:48, neurons:32)","Done","0.34", "0.91"])
ptable4.add_row(["LSTM With 2_Layer(neurons:64, neurons:48)","Done","0.32", "0.92"])

print(ptable4)





 ********** Machine Learning Model Comparison ************
+-------------------------------------------------------------------+
|                         Model Comparision                         |
+-------------------------------+------------------------+----------+
|           Model Name          | Hyperparameter Tunning | Accuracy |
+-------------------------------+------------------------+----------+
| Support Vector Machine (poly) |          Done          |  96.3%   |
|      Logistic Regression      |          Done          |  95.86%  |
|            XGBoost            |          Done          |  95.05%  |
|         Random Forest         |          Done          |  91.96%  |
|      K-Nearest Neighbours     |          Done          |  90.23%  |
|          DecisionTree         |          Done          |  86.39%  |
+-------------------------------+------------------------+----------+


 ********** Feature Importances For Best Model  ************
+-----------------------------------

1. **_Machine Learning Model Comparision_** : We can choose **_Support Vector Machine (poly)_** or **_Logistic Regression_** or **_XGBoost_** as our best model while applying ML Classical Model. XGBoost takes significantly more time than any other model to train like more that 50 times the time taken by Logistic Regression.


2. **_Dimensionality Reduction_** : If you want to reduce the dimensions of the features, Using **_PCA_** with threshold of **_0.95_** and then using **_Support Vector Machine (rbf)_** model would be best as it approximately matches the accuracy of the best model while using all the features.


3. **_Deep Learning LSTM Model Comparision_** : When we talking about LSTM Model, here with LSTM we are using simple RAW data(in ML model we are using Single engineered data made by an expert), but we can see the result without any FE data, LSTM perform very-very well and got highest 92% accuracy with 2_layer LSTM with hyperparameter Tunning and also we can clearly when we are increasing LSTM layer and Hyperparameter Tunning the cross-entropy value is decreasing and Accuracy is increasing.
